pkg/controller/feed/resources/job_test.go

```diff
@@ -0,0 +1,419 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"fmt"
+	"github.com/google/go-cmp/cmp"
+	feedsv1alpha1 "github.com/knative/eventing/pkg/apis/feeds/v1alpha1"
+	"github.com/knative/eventing/pkg/sources"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strings"
+	"testing"
+)
+
+const (
+	feedName      = "feed-name"
+	feedNamespace = "feed-namespace"
+	feedUid       = "feed-uid"
+	feedTarget    = "feed-target"
+)
+
+func TestIsJobComplete(t *testing.T) {
+	testCases := []struct {
+		name       string
+		conditions []batchv1.JobCondition
+		want       bool
+	}{
+		{
+			name: "no conditions: false",
+			want: false,
+		},
+		{
+			name: "no job complete status: false",
+			conditions: []batchv1.JobCondition{
+				{
+					Type:   batchv1.JobFailed,
+					Status: corev1.ConditionTrue,
+				},
+			},
+			want: false,
+		},
+		{
+			name: "job complete false: false",
+			conditions: []batchv1.JobCondition{
+				{
+					Type:   batchv1.JobComplete,
+					Status: corev1.ConditionFalse,
+				},
+			},
+			want: false,
+		},
+		{
+			name: "job complete true: true",
+			conditions: []batchv1.JobCondition{
+				{
+					Type:   batchv1.JobComplete,
+					Status: corev1.ConditionTrue,
+				},
+			},
+			want: true,
+		},
+		{
+			name: "multiple conditions, job complete true: true",
+			conditions: []batchv1.JobCondition{
+				{
+					Type:   batchv1.JobFailed,
+					Status: corev1.ConditionFalse,
+				},
+				{
+					Type:   batchv1.JobComplete,
+					Status: corev1.ConditionTrue,
+				},
+			},
+			want: true,
+		},
+	}
+
+	for _, tc := range testCases {
```

Comment:
You don't use `tc.name` at all.

What about combining this and the next set of cases, and using `T.Run` to make these subtests with the appropriate name:

```golang
for _, tc := range testCases {
  if !t.Run(tc.name, func(t *testing.T) {
    job := makeJobWithConditions(tc.conditions)
    got := IsJobComplete(job)
    if got != tc.want {
      t.Errorf("%s: want %v, got %v", tc.name, tc.want, got)
    }
  })
}
```

Reply:
Now that I switched to `t.Run(tc.name, ...)`, isn't this implicit? 

Forcing an error, this is what I see running `go test ./...`

```
--- FAIL: TestIsJobFailed (0.00s)
    --- FAIL: TestIsJobFailed/multiple_conditions,_job_failed_true:_true (0.00s)
        job_test.go:166: Expected false, actual true
FAIL
FAIL    github.com/knative/eventing/pkg/controller/feed/resources       0.122s
```

Which clearly includes the test's name. Is it worth adding the name a second time into my explicit error message? 

More generally, should every `t.Errorf()` include the test case's name?

Reply:
I'd suggested using `t.Run` to avoid needing to put the test case name in each `t.Errorf()`. I'd leave it as is.